File: coconut/src/mysql.rs
use std::fs::{create_dir_all, File};
use std::io::{Error as IoError, ErrorKind as IoErrorKind, Write};
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};

use clap::Parser;
use mysql::{prelude::*, Conn as DbConn, Opts as DbOpts};
use palm::{timestamp_file, Result};

#[derive(Parser, PartialEq, Eq, Debug, Clone)]
pub struct Config {
    #[arg(short, long, default_value = super::Args::DEFAULT_KEEP)]
    pub keep: usize,
    #[arg(short, long, default_value = super::Args::DEFAULT_TARGET)]
    pub target: PathBuf,
    #[arg(short = 'H', long, default_value = "127.0.0.1")]
    pub host: String,
    #[arg(short, long, default_value = "3306")]
    pub port: u16,
    #[arg(short, long)]
    pub name: String,
    #[arg(short, long, default_value = "root")]
    pub user: String,
    #[arg(short = 'P', long)]
    pub password: Option<String>,
}

impl Config {
    pub fn launch<P: AsRef<Path>>(&self, target: P) -> Result<String> {
        let target = target.as_ref();
        info!(
            "backup mysql://{}@{}:{}/{} to {}",
            self.user,
            self.host,
            self.port,
            self.name,
            target.display()
        );

        {
            let opt = DbOpts::from_url(&format!(
                "mysql://{}:{}@{}:{}/{}",
                self.user,
                self.password.as_deref().unwrap_or_default(),
                self.host,
                self.port,
                self.name
            ))?;
            let mut cli = DbConn::new(opt)?;
            let row: Option<String> = cli.query_first("select version()")?;

            info!(
                "{}",
                row.ok_or_else(|| Box::new(IoError::from(IoErrorKind::UnexpectedEof)))?
            );
        }

        let name = timestamp_file(&self.name, Some("sql"));
        let tmp = target.join(&name);
        create_dir_all(target)?;

        {
            let out = if let Some(ref password) = self.password {
                Command::new("mysqldump")
                    .arg("-h")
                    .arg(&self.host)
                    .arg("-P")
                    .arg(self.port.to_string())
                    .arg("-u")
                    .arg(&self.user)
                    .arg(format!("-p{password}"))
                    .arg(&self.name)
                    .stdout(Stdio::piped())
                    .output()?
            } else {
                Command::new("mysqldump")
                    .arg("-h")
                    .arg(&self.host)
                    .arg("-P")
                    .arg(self.port.to_string())
                    .arg("-u")
                    .arg(&self.user)
                    .arg(&self.name)
                    .stdout(Stdio::piped())
                    .output()?
            };

            if !out.status.success() {
                error!("({}) {}", out.status, std::str::from_utf8(&out.stderr)?);
                return Err(Box::new(IoError::from(IoErrorKind::UnexpectedEof)));
            }

            {
                debug!("generate {}", tmp.display());
                let mut file = File::create(&tmp)?;
                file.write_all(&out.stdout)?;
            }
        }

        info!(
            r###"USAGE:
$ unxz {file}.xz
$ mysql -h {host} -P {port} -u {user} -p {name} < {file}"###,
            file = name,
            user = self.user,
            host = self.host,
            port = self.port,
            name = self.name,
        );

        Ok(name)
    }
}

File: coconut/src/mysql.cpp
#include "coconut/mysql.hpp"
#include "coconut/nut.hpp"

coconut::MySql::MySql(const toml::table& root) {
  this->host = root["host"].value_or("127.0.0.1");
  this->port = root["port"].value_or(3306);
  this->user = root["user"].value_or("root");
  this->password = root["password"].value<std::string>();
  this->name = root["name"].value<std::string>().value();
}

void coconut::MySql::dump(const std::filesystem::path& target,
                          const size_t keep) const {
  {
    const std::string it = target.string();
    spdlog::debug("change workdir to {}", it);
    const auto ok = chdir(it.c_str());
    if (ok != EXIT_SUCCESS) {
      throw std::runtime_error("change to folder {}" + it);
    }
  }

  std::filesystem::path name(this->name + "-" + coconut::timestamp());
  name.replace_extension("sql");

  auto sql = target / name;

  spdlog::info("dump mysql://{}@{}:{}/{} to {}", this->user, this->host,
               this->port, this->name, sql.string());
  {
    std::stringstream ss;
    ss << "mysqldump"
       << " -h " << this->host << " -P " << this->port << " -u " << this->user;
    if (this->password) {
      ss << " -p" << this->password.value();
    }
    ss << " " << this->name << " > " << sql.string();
    const auto cmd = ss.str();
    spdlog::debug("{}", cmd);
    const auto ok = std::system(cmd.c_str());
    if (ok != EXIT_SUCCESS) {
      throw std::runtime_error("backup mysql");
    }
  }
  coconut::xz(sql);

  spdlog::info(R"RAW(USAGE:
$ cd {0}
$ unxz {1}.xz
$ mysql -h {2} -P {3} -u {4} -p {5} < {1} )RAW",
               target.string(), sql.string(), this->host, this->port,
               this->user, this->name);

  coconut::keep(target, keep);
}
